File: server/typescript/packages/render/test/extract/extract.test.ts
import { describe, test, expect } from "bun:test";
import { extract } from "../../src/extract/extract.js";
import {
  Format,
  FieldKind,
  FieldExtraction,
  scalar,
  enumField,
  object,
  extractSchema,
  type FieldSpec,
  type ExtractSchema,
} from "../../src/extract/types.js";

// Mirrors Java ExtractTest / C# ExtractTests (FR-010 entry-point pipeline).
function jsonAnswer(): ExtractSchema {
  return extractSchema(Format.JSON, "answer", [
    scalar("text", FieldKind.STRING, true),
    enumField("confidence", true, ["HIGH", "OK", "LOW"], { medium: "OK" }),
    scalar("note", FieldKind.STRING, false),
  ]);
}

function arrayField(name: string, kind: FieldKind, values: string[] | null, aliases: Record<string, string> | null): FieldSpec {
  return {
    name,
    kind,
    required: false,
    array: true,
    enumValues: values,
    enumAlias: aliases,
    min: null,
    max: null,
    nested: null,
    coerceDefault: null,
    defaultValue: null,
    normalize: "strip",
  };
}

describe("extract pipeline", () => {
  test("clean json all extracted", () => {
    const o = extract('{"text":"hi","confidence":"HIGH","note":"n"}', jsonAnswer());
    expect(o.data["text"]).toBe("hi");
    expect(o.data["confidence"]).toBe("HIGH");
    expect(o.report.states().get("confidence")).toBe(FieldExtraction.EXTRACTED);
    expect(o.report.hasLostRequired()).toBe(false);
  });

  test("fenced and prose-wrapped still extracts", () => {
    const dirty = 'Sure!\n```json\n{"text":"hi","confidence":"HIGH"}\n```\nDone.';
    const o = extract(dirty, jsonAnswer());
    expect(o.data["text"]).toBe("hi");
    expect(o.report.states().get("note")).toBe(FieldExtraction.LOST_OPTIONAL);
  });

  test("alias folds off-vocab", () => {
    const o = extract('{"text":"hi","confidence":"medium"}', jsonAnswer());
    expect(o.data["confidence"]).toBe("OK");
    expect(o.report.states().get("confidence")).toBe(FieldExtraction.EXTRACTED);
  });

  test("off-vocab required is malformed", () => {
    const o = extract('{"text":"hi","confidence":"banana"}', jsonAnswer());
    expect(o.report.states().get("confidence")).toBe(FieldExtraction.MALFORMED);
    expect(Object.prototype.hasOwnProperty.call(o.data, "confidence")).toBe(false);
  });

  test("missing required is lost_required", () => {
    const o = extract('{"text":"hi"}', jsonAnswer());
    expect(o.report.lostRequired()).toContain("confidence");
  });

  test("empty response flags empty and all required lost", () => {
    const o = extract("   ", jsonAnswer());
    expect(o.report.isEmpty()).toBe(true);
    expect(o.report.lostRequired()).toContain("text");
    expect(o.report.lostRequired()).toContain("confidence");
  });

  test("xml unclosed tag extracts", () => {
    const xml = extractSchema(Format.XML, "answer", [
      scalar("text", FieldKind.STRING, true),
      enumField("confidence", true, ["HIGH"], {}),
    ]);
    const o = extract("<answer><text>hi<confidence>HIGH</confidence></answer>", xml);
    expect(o.data["text"]).toBe("hi");
    expect(o.data["confidence"]).toBe("HIGH");
  });

  test("never throws on garbage", () => {
    const o = extract("@@@ totally broken @@@", jsonAnswer());
    expect(o.report.isEmpty()).toBe(true);
  });

  test("json string array extracts as list", () => {
    const s = extractSchema(Format.JSON, "answer", [arrayField("tags", FieldKind.STRING, null, null)]);
    const o = extract('{"tags":["a","b"]}', s);
    expect(o.data["tags"]).toEqual(["a", "b"]);
    expect(o.report.states().get("tags")).toBe(FieldExtraction.EXTRACTED);
  });

  test("json enum array coerces per element", () => {
    const s = extractSchema(Format.JSON, "answer", [arrayField("tones", FieldKind.ENUM, ["HIGH", "LOW"], { warm: "HIGH" })]);
    const o = extract('{"tones":["warm","LOW"]}', s);
    expect(o.data["tones"]).toEqual(["HIGH", "LOW"]);
    expect(o.report.states().get("tones")).toBe(FieldExtraction.EXTRACTED);
  });

  test("list for scalar field is malformed", () => {
    const s = extractSchema(Format.JSON, "answer", [scalar("text", FieldKind.STRING, true)]);
    const o = extract('{"text":["a","b"]}', s);
    expect(o.report.states().get("text")).toBe(FieldExtraction.MALFORMED);
    expect(Object.prototype.hasOwnProperty.call(o.data, "text")).toBe(false);
  });

  test("object field with scalar value is malformed", () => {
    const nested = extractSchema(Format.JSON, "meta", [scalar("n", FieldKind.STRING, true)]);
    const s = extractSchema(Format.JSON, "answer", [object("meta", true, false, nested)]);
    const o = extract('{"meta":"oops"}', s);
    expect(o.report.states().get("meta")).toBe(FieldExtraction.MALFORMED);
  });

  test("nested object descends and extracts", () => {
    const nested = extractSchema(Format.JSON, "meta", [scalar("n", FieldKind.STRING, true)]);
    const s = extractSchema(Format.JSON, "answer", [object("meta", true, false, nested)]);
    const o = extract('{"meta":{"n":"hello"}}', s);
    expect(o.report.states().get("meta")).toBe(FieldExtraction.EXTRACTED);
    expect(o.report.states().get("meta.n")).toBe(FieldExtraction.EXTRACTED);
    expect((o.data["meta"] as Record<string, unknown>)["n"]).toBe("hello");
  });

  test("truncated value is malformed not lost", () => {
    const o = extract('{"text":"hi","confidence":', jsonAnswer());
    expect(o.data["text"]).toBe("hi");
    expect(o.report.states().get("confidence")).toBe(FieldExtraction.MALFORMED);
    expect(o.report.isEmpty()).toBe(false);
  });

  test("partial enum array is malformed but keeps valid elements", () => {
    const s = extractSchema(Format.JSON, "answer", [arrayField("tones", FieldKind.ENUM, ["HIGH", "LOW"], {})]);
    const o = extract('{"tones":["HIGH","grape"]}', s);
    expect(o.report.states().get("tones")).toBe(FieldExtraction.MALFORMED);
    expect(o.data["tones"]).toEqual(["HIGH"]);
  });

  // ---- FR-011 DEFAULTED classification + @default absent-fill ----
  test("present-garbage enum with coerceDefault is DEFAULTED and satisfies required", () => {
    const s = extractSchema(Format.JSON, "answer", [
      enumField("confidence", true, ["HIGH", "LOW"], {}, "LOW"),
    ]);
    const o = extract('{"confidence":"banana"}', s);
    expect(o.data["confidence"]).toBe("LOW");
    expect(o.report.states().get("confidence")).toBe(FieldExtraction.DEFAULTED);
    expect(o.report.lostRequired()).not.toContain("confidence");
  });

  test("present-valid enum stays EXTRACTED (not DEFAULTED)", () => {
    const s = extractSchema(Format.JSON, "answer", [
      enumField("confidence", true, ["HIGH", "LOW"], {}, "LOW"),
    ]);
    const o = extract('{"confidence":"HIGH"}', s);
    expect(o.data["confidence"]).toBe("HIGH");
    expect(o.report.states().get("confidence")).toBe(FieldExtraction.EXTRACTED);
  });

  test("absent enum with @default is DEFAULTED and satisfies required", () => {
    const s = extractSchema(Format.JSON, "answer", [
      enumField("confidence", true, ["HIGH", "LOW"], {}, null, "strip", "HIGH"),
    ]);
    const o = extract("{}", s);
    expect(o.data["confidence"]).toBe("HIGH");
    expect(o.report.states().get("confidence")).toBe(FieldExtraction.DEFAULTED);
    expect(o.report.lostRequired()).not.toContain("confidence");
    expect(o.report.coercions().some((c) => c.fieldPath === "confidence" && c.kind === "default")).toBe(true);
  });

  test("absent enum without @default is LOST_REQUIRED", () => {
    const s = extractSchema(Format.JSON, "answer", [
      enumField("confidence", true, ["HIGH", "LOW"], {}, null),
    ]);
    const o = extract("{}", s);
    expect(o.report.lostRequired()).toContain("confidence");
    expect(Object.prototype.hasOwnProperty.call(o.data, "confidence")).toBe(false);
  });

  test("absent optional enum with @default fills DEFAULTED", () => {
    const s = extractSchema(Format.JSON, "answer", [
      enumField("confidence", false, ["HIGH", "LOW"], {}, null, "strip", "LOW"),
    ]);
    const o = extract("{}", s);
    expect(o.data["confidence"]).toBe("LOW");
    expect(o.report.states().get("confidence")).toBe(FieldExtraction.DEFAULTED);
  });

  test("nested object extracts with dotted child path EXTRACTED", () => {
    const nested = extractSchema(Format.JSON, "meta", [scalar("score", FieldKind.INT, true)]);
    const s = extractSchema(Format.JSON, "answer", [object("meta", true, false, nested)]);
    const o = extract('{"meta":{"score":7}}', s);
    expect(o.report.states().get("meta")).toBe(FieldExtraction.EXTRACTED);
    expect(o.report.states().get("meta.score")).toBe(FieldExtraction.EXTRACTED);
    expect((o.data["meta"] as Record<string, unknown>)["score"]).toBe(7);
  });
});

File: server/java/render/src/test/java/com/metaobjects/render/extract/ExtractTest.java
package com.metaobjects.render.extract;

import org.junit.Test;
import java.util.List;
import java.util.Map;
import static org.junit.Assert.*;

public class ExtractTest {

    private ExtractSchema jsonAnswer() {
        return new ExtractSchema(Format.JSON, "answer", List.of(
                FieldSpec.scalar("text", FieldKind.STRING, true),
                FieldSpec.enumField("confidence", true, List.of("HIGH", "OK", "LOW"), Map.of("medium", "OK")),
                FieldSpec.scalar("note", FieldKind.STRING, false)));
    }

    @Test
    public void cleanJsonAllExtracted() {
        ExtractionOutcome o = Extract.extract(
                "{\"text\":\"hi\",\"confidence\":\"HIGH\",\"note\":\"n\"}", jsonAnswer(), ExtractOptions.defaults());
        assertEquals("hi", o.data().get("text"));
        assertEquals("HIGH", o.data().get("confidence"));
        assertEquals(FieldExtraction.EXTRACTED, o.report().states().get("confidence"));
        assertFalse(o.report().hasLostRequired());
    }

    @Test
    public void fencedAndProseWrappedStillExtracts() {
        String dirty = "Sure!\n```json\n{\"text\":\"hi\",\"confidence\":\"HIGH\"}\n```\nDone.";
        ExtractionOutcome o = Extract.extract(dirty, jsonAnswer(), ExtractOptions.defaults());
        assertEquals("hi", o.data().get("text"));
        assertEquals(FieldExtraction.LOST_OPTIONAL, o.report().states().get("note"));
    }

    @Test
    public void aliasFoldsOffVocab() {
        ExtractionOutcome o = Extract.extract(
                "{\"text\":\"hi\",\"confidence\":\"medium\"}", jsonAnswer(), ExtractOptions.defaults());
        assertEquals("OK", o.data().get("confidence"));
        assertEquals(FieldExtraction.EXTRACTED, o.report().states().get("confidence"));
    }

    @Test
    public void offVocabRequiredIsMalformed() {
        ExtractionOutcome o = Extract.extract(
                "{\"text\":\"hi\",\"confidence\":\"banana\"}", jsonAnswer(), ExtractOptions.defaults());
        assertEquals(FieldExtraction.MALFORMED, o.report().states().get("confidence"));
        assertFalse(o.data().containsKey("confidence"));
    }

    @Test
    public void missingRequiredIsLostRequired() {
        ExtractionOutcome o = Extract.extract("{\"text\":\"hi\"}", jsonAnswer(), ExtractOptions.defaults());
        assertEquals(List.of("confidence"), o.report().lostRequired());
    }

    @Test
    public void emptyResponseFlagsEmptyAndAllRequiredLost() {
        ExtractionOutcome o = Extract.extract("   ", jsonAnswer(), ExtractOptions.defaults());
        assertTrue(o.report().isEmpty());
        assertTrue(o.report().lostRequired().contains("text"));
        assertTrue(o.report().lostRequired().contains("confidence"));
    }

    @Test
    public void xmlUnclosedTagExtracts() {
        ExtractSchema xml = new ExtractSchema(Format.XML, "answer", List.of(
                FieldSpec.scalar("text", FieldKind.STRING, true),
                FieldSpec.enumField("confidence", true, List.of("HIGH"), Map.of())));
        ExtractionOutcome o = Extract.extract("<answer><text>hi<confidence>HIGH</confidence></answer>",
                xml, ExtractOptions.defaults());
        assertEquals("hi", o.data().get("text"));
        assertEquals("HIGH", o.data().get("confidence"));
    }

    @Test
    public void neverThrowsOnGarbage() {
        ExtractionOutcome o = Extract.extract("@@@ totally broken @@@", jsonAnswer(), ExtractOptions.defaults());
        assertTrue(o.report().isEmpty());
    }

    @Test
    public void jsonStringArrayExtractsAsList() {
        ExtractSchema s = new ExtractSchema(Format.JSON, "answer", List.of(
                new FieldSpec("tags", FieldKind.STRING, false, true, null, null, null, null, null,
                        null, null, Normalize.DEFAULT, false)));
        ExtractionOutcome o = Extract.extract("{\"tags\":[\"a\",\"b\"]}", s, ExtractOptions.defaults());
        assertEquals(List.of("a", "b"), o.data().get("tags"));
        assertEquals(FieldExtraction.EXTRACTED, o.report().states().get("tags"));
    }

    @Test
    public void jsonEnumArrayCoercesPerElement() {
        ExtractSchema s = new ExtractSchema(Format.JSON, "answer", List.of(
                new FieldSpec("tones", FieldKind.ENUM, false, true,
                        List.of("HIGH", "LOW"), Map.of("warm", "HIGH"), null, null, null,
                        null, null, Normalize.DEFAULT, false)));
        ExtractionOutcome o = Extract.extract("{\"tones\":[\"warm\",\"LOW\"]}", s, ExtractOptions.defaults());
        assertEquals(List.of("HIGH", "LOW"), o.data().get("tones"));
        assertEquals(FieldExtraction.EXTRACTED, o.report().states().get("tones"));
    }

    @Test
    public void listForScalarFieldIsMalformed() {
        ExtractSchema s = new ExtractSchema(Format.JSON, "answer", List.of(
                FieldSpec.scalar("text", FieldKind.STRING, true)));
        ExtractionOutcome o = Extract.extract("{\"text\":[\"a\",\"b\"]}", s, ExtractOptions.defaults());
        assertEquals(FieldExtraction.MALFORMED, o.report().states().get("text"));
        assertFalse(o.data().containsKey("text"));
    }

    @Test
    public void objectFieldWithScalarValueIsMalformed() {
        ExtractSchema nested = new ExtractSchema(Format.JSON, "meta",
                List.of(FieldSpec.scalar("n", FieldKind.STRING, true)));
        ExtractSchema s = new ExtractSchema(Format.JSON, "answer", List.of(
                FieldSpec.object("meta", true, false, nested)));
        ExtractionOutcome o = Extract.extract("{\"meta\":\"oops\"}", s, ExtractOptions.defaults());
        assertEquals(FieldExtraction.MALFORMED, o.report().states().get("meta"));
    }

    @Test
    public void truncatedValueIsMalformedNotLost() {
        // confidence key present but value cut off → MALFORMED (present-but-garbled), distinct from absent
        ExtractionOutcome o = Extract.extract("{\"text\":\"hi\",\"confidence\":", jsonAnswer(), ExtractOptions.defaults());
        assertEquals("hi", o.data().get("text"));
        assertEquals(FieldExtraction.MALFORMED, o.report().states().get("confidence"));
        assertFalse(o.report().isEmpty());
    }

    @Test
    public void partialEnumArrayIsMalformedButKeepsValidElements() {
        ExtractSchema s = new ExtractSchema(Format.JSON, "answer", List.of(
                new FieldSpec("tones", FieldKind.ENUM, false, true,
                        List.of("HIGH", "LOW"), Map.of(), null, null, null,
                        null, null, Normalize.DEFAULT, false)));
        ExtractionOutcome o = Extract.extract("{\"tones\":[\"HIGH\",\"grape\"]}", s, ExtractOptions.defaults());
        assertEquals(FieldExtraction.MALFORMED, o.report().states().get("tones"));
        assertEquals(List.of("HIGH"), o.data().get("tones"));   // valid element retained
    }
}
